File: apps/cargo-tracker/src/Cargotracker/Booking/Domain/Model/Leg.hs
{- | Leg エンティティ (US09, IT4)

経路 (Itinerary) を構成する 1 区間。
ADR-0004 Cross-BC 規約に従い、`voyageNumber` / `loadLocation` / `unloadLocation` は
Text として保持し、Routing BC の `VoyageNumber` / Shared.UnLocode を直接 import しない。
-}
module Cargotracker.Booking.Domain.Model.Leg
  ( Leg (..),
    mkLeg,
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import Data.Time (UTCTime)

import Cargotracker.Shared.Domain.DomainError (DomainError (..))

data Leg = Leg
  { legSeqNumber :: !Int
  , legLoadLocation :: !Text
  -- ^ UnLocode 文字列 (5 文字、ADR-0004)
  , legUnloadLocation :: !Text
  -- ^ UnLocode 文字列 (5 文字、ADR-0004)
  , legLoadTime :: !UTCTime
  , legUnloadTime :: !UTCTime
  , legVoyageNumber :: !Text
  -- ^ VoyageNumber 文字列 (ADR-0004)
  }
  deriving stock (Eq, Show)

-- | スマートコンストラクタ。順序・空文字を検証する。
mkLeg ::
  -- | seq_number (1 以上)
  Int ->
  -- | load_location_unlocode
  Text ->
  -- | unload_location_unlocode
  Text ->
  -- | load_time
  UTCTime ->
  -- | unload_time
  UTCTime ->
  -- | voyage_number
  Text ->
  Either DomainError Leg
mkLeg seq_ loadLoc unloadLoc loadT unloadT voyNum
  | seq_ < 1 = Left (InvalidLeg "seq_number must be >= 1")
  | T.null loadLoc = Left (InvalidLeg "load_location is empty")
  | T.null unloadLoc = Left (InvalidLeg "unload_location is empty")
  | T.null voyNum = Left (InvalidLeg "voyage_number is empty")
  | loadT >= unloadT = Left (InvalidLeg "load_time must be < unload_time")
  | otherwise =
      Right
        Leg
          { legSeqNumber = seq_
          , legLoadLocation = loadLoc
          , legUnloadLocation = unloadLoc
          , legLoadTime = loadT
          , legUnloadTime = unloadT
          , legVoyageNumber = voyNum
          }

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/booking/domain/model/valueobjects/Leg.java
package com.example.cargotracker.booking.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.model.valueobjects.Location;
import java.time.Instant;

/**
 * 輸送区間。単一航海での積込港から荷降港までの区間。
 *
 * <p>航海番号は<strong>文字列で持つ</strong>。Routing の {@code VoyageNumber} を
 * 参照すると BC 間の直接参照になる（ADR-005・ArchUnit ルール 4）。
 * 貨物の側が知る必要があるのは「どの便で運ぶか」という事実だけである。
 *
 * @param voyageNumber   航海番号
 * @param loadLocation   積込港
 * @param unloadLocation 荷降港
 * @param loadTime       積込予定日時
 * @param unloadTime     荷降予定日時
 */
public record Leg(
        String voyageNumber,
        Location loadLocation,
        Location unloadLocation,
        Instant loadTime,
        Instant unloadTime) {

    public Leg {
        if (voyageNumber == null || voyageNumber.isBlank()) {
            // どの便で運ぶかが分からない区間は、荷役の照合にも使えない
            throw new IllegalArgumentException("航海番号は必須です");
        }
        if (loadLocation == null || unloadLocation == null) {
            throw new IllegalArgumentException("積込港と荷降港は必須です");
        }
        if (loadLocation.equals(unloadLocation)) {
            throw new IllegalArgumentException(
                    "積込港と荷降港が同じです: " + loadLocation.unlocode());
        }
        if (loadTime == null || unloadTime == null) {
            throw new IllegalArgumentException("積込日時と荷降日時は必須です");
        }
        if (!unloadTime.isAfter(loadTime)) {
            throw new IllegalArgumentException("荷降日時は積込日時より後です");
        }
        voyageNumber = voyageNumber.strip();
    }

    public static Leg of(
            String voyageNumber,
            Location loadLocation,
            Location unloadLocation,
            Instant loadTime,
            Instant unloadTime) {
        return new Leg(voyageNumber, loadLocation, unloadLocation, loadTime, unloadTime);
    }
}
